### PR TITLE
Add minizip-ng 3.0.1

### LINKED ISSUE #4386: [request] minizip-ng/3.0.0
### Package Details
  * Package Name/Version: **minizip-ng/3.0.0**
  * Website: **https://github.com/zlib-ng/minizip-ng**
  * Source code: **https://github.com/zlib-ng/minizip-ng**

### Description Of The Library / Tool
Fork of minizip (cross-platform zip manipulation library written in C) which already has recipe here with bugfixes and new features.

It has 738 stars currently.

### Additional information
Optional dependencies:
- bzip2
- iconv
- libbsd
- liblzma
- openssl
- zlib
- zstd

Probably some changes should be made in CMakeLists.txt for easier packaging and proper injection of dependencies

### PR DESCRIPTION
**Pull request by uilianries**
Specify library name and version:  **minizip-ng/3.0.0**

PR #4394 is stale, but almost good to be merged. This PR applies review suggestions.

closes #4386
closes #4394 (supersede)

---

- [ ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/minizip-ng/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        if self.options.get_safe("with_zlib"):
+            self.requires("zlib/1.2.11")
+        if self.options.with_bzip2:
+            self.requires("bzip2/1.0.8")
+        if self.options.with_lzma:
+            self.requires("xz_utils/5.2.5")
+        if self.options.with_zstd:
+            self.requires("zstd/1.4.8")
+        if self.options.with_openssl:
+            self.requires("openssl/1.1.1i")
```
Comment (prince-chrismc) [inline recipes/minizip-ng/all/conanfile.py:84]: There should be a k that's released

**Review by prince-chrismc — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

File: recipes/minizip-ng/all/conanfile.py (hunk 1 above)
Comment (SpaceIm) [reply recipes/minizip-ng/all/conanfile.py:84]: yes

File: recipes/minizip-ng/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+        del self.settings.compiler.cppstd
+        del self.settings.compiler.libcxx
+        if self.options.mz_compatibility:
+            self.provides = "minizip"
+        if not tools.is_apple_os(self.settings.os):
+            del self.options.with_libcomp
```
Comment (SpaceIm) [inline recipes/minizip-ng/all/conanfile.py:70]: this condition should go in `config_options()` I think

File: recipes/minizip-ng/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+            self.requires("bzip2/1.0.8")
+        if self.options.with_lzma:
+            self.requires("xz_utils/5.2.5")
+        if self.options.with_zstd:
+            self.requires("zstd/1.4.8")
+        if self.options.with_openssl:
+            self.requires("openssl/1.1.1i")
+        if self.settings.os != "Windows":
+            if self.options.with_iconv:
+                self.requires("libiconv/1.16")
```
Comment (SpaceIm) [inline recipes/minizip-ng/all/conanfile.py:87]:
```suggestion
        if self.options.get_safe("with_iconv"):
            self.requires("libiconv/1.16")
```

File: recipes/minizip-ng/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+                              "BZip2_FOUND")
+        tools.replace_in_file(os.path.join(self._source_subfolder, "CMakeLists.txt"),
+                              "BZIP2_VERSION_STRING",
+                              "BZip2_VERSION")
+        tools.replace_in_file(os.path.join(self._source_subfolder, "CMakeLists.txt"),
+                              "BZIP2_INCLUDE_DIRS",
+                              "BZip2_INCLUDE_DIRS")
+        tools.replace_in_file(os.path.join(self._source_subfolder, "CMakeLists.txt"),
+                              "BZIP2_LIBRARIES",
+                              "BZip2_LIBRARIES")
```
Comment (SpaceIm) [inline recipes/minizip-ng/all/conanfile.py:126]: not required anymore, bzip2 recipe now provides those variables in `cmake_find_package` generator.

File: recipes/minizip-ng/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        tools.replace_in_file(os.path.join(self._source_subfolder, "CMakeLists.txt"),
+                              "BZIP2_LIBRARY_DIRS",
+                              "BZip2_LIB_DIRS")
```
Comment (SpaceIm) [inline recipes/minizip-ng/all/conanfile.py:129]: probably not required, but I'm not sure

**Comment by SpaceIm** [comment]
fPIC option is not honored: https://github.com/zlib-ng/minizip-ng/blob/1593060ecf093411f1f20f655e50992913c6d5b9/CMakeLists.txt#L709-L711

File: recipes/minizip-ng/all/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.4)
+project(cmake_wrapper)
+
+include(conanbuildinfo.cmake)
+conan_basic_setup()
+
+set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
```
Comment (SpaceIm) [inline recipes/minizip-ng/all/CMakeLists.txt:7]: Are you sure it's required? symbols should be properly exported: https://github.com/zlib-ng/minizip-ng/blob/1593060ecf093411f1f20f655e50992913c6d5b9/mz.h#L150-L154

File: recipes/minizip-ng/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    name = "minizip-ng"
+    description = "Fork of the popular zip manipulation library found in the zlib distribution."
+    topics = ("conan", "compression", "zip")
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/zlib-ng/minizip-ng"
+    license = "Zlib"
+    exports_sources = "CMakeLists.txt"
+    generators = "cmake", "cmake_find_package", "pkg_config"
+    settings = "os", "compiler", "build_type", "arch"
+    provides = "minizip"
```
Comment (SpaceIm) [inline recipes/minizip-ng/all/conanfile.py:15]:
```suggestion
```

conditionally set below in `configure()`

File: recipes/minizip-ng/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        return "build_subfolder"
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+            del self.options.with_iconv
+            del self.options.with_libbsd
+
+    def build_requirements(self):
+        self.build_requires('pkgconf/1.7.3')
```
Comment (SpaceIm) [inline recipes/minizip-ng/all/conanfile.py:60]: With a small fix in `xz_utils` recipe, I think that this build requirement could be removed.
But `pkg_config` generator should be kept. Indeed, minizip-ng tries first to find dependencies through pkg-config -but only for lzma, openssl and zstd- then fall back to CMake find/config files.

File: recipes/minizip-ng/all/CMakeLists.txt (hunk 6 above)
Comment (madebr) [reply recipes/minizip-ng/all/CMakeLists.txt:7]: There's no `__declspec(dllimport)`.
Is this a private or public header?

Comment (SpaceIm) [reply recipes/minizip-ng/all/CMakeLists.txt:7]: __declspec(dllimport) is not mandatory, except for global symbols.

File: recipes/minizip-ng/all/conanfile.py (hunk 1 above)
Comment (uilianries) [reply recipes/minizip-ng/all/conanfile.py]: done

File: recipes/minizip-ng/all/conanfile.py (hunk 3 above)
Comment (uilianries) [reply recipes/minizip-ng/all/conanfile.py]: done

File: recipes/minizip-ng/all/conanfile.py (hunk 7 above)
Comment (uilianries) [reply recipes/minizip-ng/all/conanfile.py]: done

File: recipes/minizip-ng/all/conanfile.py (hunk 5 above)
Comment (uilianries) [reply recipes/minizip-ng/all/conanfile.py]: removed, let's see if it works

File: recipes/minizip-ng/all/conanfile.py (hunk 4 above)
Comment (uilianries) [reply recipes/minizip-ng/all/conanfile.py]: done

File: recipes/minizip-ng/all/conanfile.py (hunk 2 above)
Comment (uilianries) [reply recipes/minizip-ng/all/conanfile.py]: done

File: recipes/minizip-ng/all/CMakeLists.txt (hunk 6 above)
Comment (uilianries) [reply recipes/minizip-ng/all/CMakeLists.txt:7]: It does not work. Only exports DLL file.

Comment (SpaceIm) [reply recipes/minizip-ng/all/CMakeLists.txt:7]: _WINDOWS macro should be replaced by _WIN32, that's the reason I think.

Comment (SpaceIm) [reply recipes/minizip-ng/all/CMakeLists.txt:7]: Ok, MZ_EXPORT is not put in front of many functions of the API, so go with CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS

Comment (uilianries) [reply recipes/minizip-ng/all/CMakeLists.txt:7]: Yes, I saw it requires more than a simple change. I would prefer sending it to upstream instead.

**Review by walterbrebels — APPROVED** [review]

**Comment by walterbrebels** [comment]
Bump! Are any additional changes needed for this PR?

**Comment by uilianries** [comment]
No, only more reviews!

**Review by SSE4 — APPROVED** [review]

**Review by madebr — APPROVED** [review]

**Review by SpaceIm — APPROVED** [review]
zstd version could be update to 1.4.9, but it's fine.

**Comment by walterbrebels** [comment]
Thank you all!

**Pull request merged**